File: Http/Metadata/Metadata.cs
using System.Collections.Generic;

namespace WebCommons.Http
{
    /// <summary>
    /// Defines the HTML page metadata for search engines.
    /// </summary>
    public class SearchEngineMetadata
    {
        public string? Title { get; set; } = null;
        public string? TitlePrefix { get; set; } = null;
        public string? TitleSuffix { get; set; } = null;
        public string FullTitle { get => string.Format("{0}{1}{2}", this.TitlePrefix, this.Title, this.TitleSuffix).Trim(); }

        public string? Section { get; set; } = null;
        public string? Description { get; set; } = null;
        public string? Image { get; set; } = null;
        public string? SiteName { get; set; } = null;
        public string? Domain { get; set; } = null;
        public string? Url { get; set; } = null;
        public string? Locale { get; set; } = "en_US";
        public DateTime? ModifiedTime { get; set; } = null;
        public RobotIndexing Robots { get; set; } = RobotIndexing.Follow;
        public PageType Type { get; set; } = PageType.Website;

        #region Enums

        public enum RobotIndexing { Follow, NoFollow }
        private static readonly Dictionary<RobotIndexing, string> RobotsMap = new() {
            { RobotIndexing.Follow, "index, follow" },
            { RobotIndexing.NoFollow, "noindex, nofollow" } };

        public enum PageType { Article, Profile, Website, Other }

        #endregion

        public Dictionary<string, string> ToMap()
        {
            Dictionary<string, string> map = new();
            map["title"] = this.Title;
            map["fullTitle"] = this.FullTitle;
            map["titlePrefix"] = this.TitlePrefix;
            map["titleSuffix"] = this.TitleSuffix;
            map["description"] = this.Description;
            map["type"] = (this.Type == PageType.Other) ? null : Enum.GetName(typeof(PageType), this.Type).ToLower();
            map["section"] = this.Section;
            map["siteName"] = this.SiteName;
            map["locale"] = this.Locale;
            map["modifiedTime"] = this.ModifiedTime?.ToString("yyyy-MM-dd");
            map["image"] = this.Domain + this.Image;
            map["robots"] = RobotsMap[this.Robots];
            map["url"] = this.Domain + this.Url;
            return map;
        }

        public string ToHtml()
        {
            var map = this.ToMap();
            return "<title>" + this.FullTitle + "</title>"
                + CreateMetadataElement("og:title", "property", this.FullTitle)
                + CreateMetadataElement("description", "name", map["description"])
                + CreateMetadataElement("og:description", "property", map["description"])
                + CreateMetadataElement("og:type", "property", map["type"])
                + CreateMetadataElement("og:section", "property", map["section"])
                + CreateMetadataElement("og:site_name", "property", map["siteName"])
                + CreateMetadataElement("og:locale", "property", map["locale"])
                + CreateMetadataElement("og:modified_time", "property", map["modifiedTime"])
                + CreateMetadataElement("og:image", "property", map["image"])
                + CreateMetadataElement("robots", "name", map["robots"])
                + CreateMetadataElement("url", "name", map["url"]);
        }

        private static string CreateMetadataElement(string name, string attribute, string value)
        {
            if (string.IsNullOrEmpty(value)) return string.Empty;
            else return string.Format("<meta {0}=\"{1}\" content=\"{2}\" />", attribute, name, value);
        }
    }
}

File: Http/Metadata/Metadata.js
/**
 * Defines the HTML page metadata for search engines.
 */
class SearchEngineMetadata {
    constructor(defaults) {
        if (typeof defaults !== 'undefined') { this.defaults = defaults; }

        this.description = null;
        this.image = null;
        this.title = null;
        this.titlePrefix = null;
        this.titleSuffix = null;
        this.robots = SearchEngineMetadata.robots.follow;
        this.type = SearchEngineMetadata.type.website;
        this.modifiedTime = null;
        this.section = null;
        this.siteName = null;
        this.locale = 'en_US';
    }

    static type = {
        article: 'article',
        profile: 'profile',
        website: 'website'
    }

    static robots = {
        follow: 'index, follow',
        nofollow: 'noindex, nofollow'
    }

    /**
     * Resets the metadata and sets the new values.
     * @param {object} values
     */
    set(values) {
        this.reset();
        Object.assign(this, values);
    }

    /**
     * Resets the values to their defaults.
     */
    reset() {
        if (typeof this.defaults === 'undefined') return;
        Object.assign(this, this.defaults);
    }

    /**
     * Creates, updates or removes metadata nodes.
     * @param {object} [values] Optional values to reset the metadata and set the new values.
     */
    apply(values) {
        if (typeof values === 'string') this.set({ title: values });
        else if (typeof values === 'object') this.set(values);

        document.title = [this.titlePrefix, this.title, this.titleSuffix].filterEmpty().join(' ').trim();
        let route = App.state.route;

        let metadata = [
            { name: 'og:title',         attribute: 'property',  value: document.title },
            { name: 'robots',           attribute: 'name',      value: this.robots },
            { name: 'description',      attribute: 'name',      value: this.description },
            { name: 'og:description',   attribute: 'property',  value: this.description },
            { name: 'og:type',          attribute: 'property',  value: this.type },
            { name: 'og:section',       attribute: 'property',  value: this.section },
            { name: 'og:modified_time', attribute: 'property',  value: this.modifiedTime },
            { name: 'og:image',         attribute: 'property',  value: this.image },
            { name: 'og:url',           attribute: 'property',  value: (route ? route.getCanonicalUri(App.state.params) : null) }
        ];

        document.head.setMetadata(metadata);
    }
}

document.metadata = new SearchEngineMetadata(new SearchEngineMetadata());